[RF-DOCS][ci-skip] Add Product Reviews Tutorial (#57244)
rails/rails: guides/source/product_reviews.md

**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON <https://guides.rubyonrails.org>.**

Product Reviews
===============

This guide covers adding Reviews to the e-commerce application you created in
the [Getting Started Guide](getting_started.html)). We will use the code from
the [Wishlists Guide](wishlists.html) as a starting place.

After reading this guide, you will know how to:

* Collect reviews
* Calculate average ratings for products
* Filter reviews

--------------------------------------------------------------------------------

Introduction
------------

> An e-commerce stores isn't complete these days without product reviews. In this

Reviewer: `An e-commerce store`

Author: ```suggestion
An e-commerce store isn't complete these days without product reviews. In this
```

guide, we'll collect reviews for products, average the ratings, and give
customers and admins a way to view and filter the reviews.

Let's get started!

Reviews Model
-------------

Product reviews typically consist of a 1 to 5 star rating and some text the user
wrote about the product.

Let's start by creating the `Review` model to store this data.

```bash
$ bin/rails generate model Review product:belongs_to user:belongs_to rating:integer body:text images:attachments
```

This model has several attributes and associations:

- `product:belongs_to` which associates the `Review` with the `Product`
- `user:belongs_to` which associates the `Review` with the `User` who created it
- `rating` which is an integer that stores the 1-5 star rating

> - `body` which stores the text description of the review

Reviewer 2: ```suggestion
- `product:belongs_to` associates the `Review` with the `Product`
- `user:belongs_to` associates the `Review` with the `User` who created it
- `rating` is an integer that stores the 1-5 star rating
- `body` stores the text description of the review
```
reads a bit faster, and matches the last attribute bullet.

- `images` stores images using ActiveStorage

### Caches

Before we run this migration, let's modify it to add a couple things to the
`products` table that we'll need.

1. A counter cache to keep track of a product's total number of reviews
2. A rating column to store the product's average rating

Open `db/migrate/<timestamp>_create_reviews.rb` and add these columns:

```ruby#12-13
class CreateReviews < ActiveRecord::Migration[8.0]
  def change
    create_table :reviews do |t|
      t.belongs_to :product, null: false, foreign_key: true
      t.belongs_to :user, null: false, foreign_key: true
      t.integer :rating

>       t.text :body

Reviewer 2: ```suggestion
      t.integer :rating, null: false
      t.text :body, null: false
```
Since we require them in the model using validations.

validator ensures that ratings are integers between 1 and 5.

### Associations

We also need to add associations to the `Product` and `User` models for reviews.

In `app/models/user.rb`, add the association:

```ruby#4
class User < ApplicationRecord
  has_secure_password
  has_many :sessions, dependent: :destroy
  has_many :reviews, dependent: :destroy
  has_many :wishlists, dependent: :destroy
```

Do the same in `app/models/product.rb`:

```ruby#4
class Product < ApplicationRecord
  include Notifications

  has_many :reviews, dependent: :destroy
  has_many :subscribers, dependent: :destroy
```

Now we're ready to start collecting product reviews!

## Collecting Reviews

On the product's show page, we can ask customers to write a review and display
all the current reviews. Let's start by letting users create reviews.

### Public product reviews routes

Let's create the routes for product reviews first. Since this is public-facing,
we only need the `new` and `create` actions.

In `config/routes.rb`, add the following inside the `resources :products` block:

```ruby#3
  resources :products do
    resource :wishlist, only: [ :create ], module: :products

>     resources :reviews, only: [ :new, :create], module: :products

Reviewer 3: ```suggestion
    resources :reviews, only: [ :new, :create ], module: :products
```

</section>
```

At the bottom of `app/views/products/show.html.erb`, we can render the partial:

```erb
<%= render "reviews", product: @product %>
```

We're passing in the product as context here so the partial knows which product
to render reviews for.

### Add the Reviews Controller

To render the form, we need to create the controller.

Create `app/controllers/products/reviews_controller.rb` with the following:

```ruby
class Products::ReviewsController < ApplicationController
  before_action :set_product

  def new
    @review = Review.new
  end

  private

  def set_product
    @product = Product.find(params[:product_id])
  end
end
```

This controller uses the nested route to look up the `Product` on each request.

### New review form

Next, let's add the view for collecting reviews.

Create `app/views/products/reviews/new.html.erb` with the following:

```erb
<h1>Add a review</h1>

<%= form_with model: [@product, @review] do |form| %>
  <div>
    <%= form.label :rating %>
    <div class="rating">
      <% 5.downto(1).each do |i| %>
        <%= form.radio_button :rating, i, required: true %>
        <%= form.label :rating, "★", value: i %>
      <% end %>
    </div>

>   </div>

Reviewer 2: ```suggestion
<fieldset>
    <legend>Rating</legend>
    <div class="rating">
      <% 1.upto(5).each do |i| %>
        <%= form.radio_button :rating, i, required: true, class: "sr-only" %>
        <%= form.label :rating, value: i do %>
          <span aria-hidden="true">★</span>
          <span class="sr-only"><%= pluralize(i, "star") %></span>
        <% end %>
      <% end %>
    </div>
  </fieldset>
```

`Fieldset/legend` group the radio inputs for screen readers, `aria-hidden` hides the star glyph (which is announced inconsistently by screen readers), and the `.sr-only` text gives meaningful input name for screen readers. Final audio rendering: `Rating, 1 star, radio button, 1 of 5`.

could choose.

For image uploads, we're using a file field with a couple attributes that do the
following:

- `multiple: true` tells the browser to allow multiple file uploads
- `accept: "image/*"` is a browser hint that filters the file selector to only
images
- `capture: "environment"` is used by mobile browsers to enable the camera using
the rear camera

### Styling the star rating input

The stars should be gray by default and turn gold when you select a rating. We
can use a few CSS tricks to make this work like you'd expect.

Add the following to `app/assets/stylesheets/application.css` at the bottom:

```css
.rating {
  display: flex;
  flex-direction: row-reverse; /* Flips the 5-1 DOM order to 1-5 visually */
  justify-content: flex-end;   /* Aligns the 1-star to the left */
}

/* Hide the radio buttons */
.rating input {
  display: none;
}

/* Lighten the unselected stars */
.gray,
.rating label {
  color: lightgray;
}

/* Target all stars after the checked/hovered one in the DOM (which appear before it visually) */
.gold,
.rating input:checked ~ label,
.rating label:hover,
.rating label:hover ~ label {
  color: gold;

> }

Reviewer 2: ```suggestion
/* Remove default styling for fieldset */
fieldset {
  border: 0;
  padding: 0;
  margin: 0;
}

/* Remove default styling for legend */
legend {
  padding: 0;
}

/* Hide text visually but not from screen readers */
.sr-only {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}

.rating {
  display: flex;
}

/* Lighten all stars by default */
.rating label {
  color: lightgray;
}

/* Gold: the selected/focused/hovered star and all previous ones */
.rating input:is(:checked, :focus) + label,
.rating label:has(~ input:is(:checked, :focus)),
.rating label:hover,
.rating label:has(~ input + label:hover) {
  color: gold;
}

/* Visible focus indicator */
.rating input:focus-visible + label {
  outline: .2em solid;
  outline-offset: 2px;
}
```

Remove fieldset/legend styles, use CSS to hide the inputs while keeping them keyboard-accessible, hover/focus styles to make previous stars gold too, and display a focus indicator for keyboard users only.

```

There are a few things going on here:

1. The ratings are rendered in reverse order so users see 1-5 stars in order
2. The radio buttons are hidden
3. When a radio is checked or hovered, the labels after it are colored gold

CSS allows us to select subsequent siblings using the tilde `~`. By using this
with reverse ordering, we can highlight all the correct stars when the user
selects a rating. If they choose 3 stars, the selector colors the stars 3, 2,

> and 1 with gold.

Reviewer 2: ```suggestion
1. The ratings display as stars, but screen reader users hear clear indications like `Rating, 1 star, radio button, 1 of 5`
2. The radio buttons are hidden but keyboard accessible
3. When a radio is checked, focused or hovered, the star labels before it are colored gold
 
CSS allows us to select previous siblings using `:has(~ input:checked)` and `:has(~ input + label:hover)`. This selector targets labels for all the lower stars when the user selects a rating. If they choose 3 stars, the selector colors the stars 1, 2, and 3 with gold.
```

A bit less readable, but it does the job!

Using `:focus-visible` prevents the outline from appearing on mouse clicks.

### Creating reviews

Next, we need to save reviews in the database when the form is submitted.

Add the `create` action to the controller:

```ruby#8-15,23-25
class Products::ReviewsController < ApplicationController
  before_action :set_product

  def new
    @review = Review.new
  end

  def create
    @review = @product.reviews.new(review_params)
    if @review.save
      redirect_to @product, notice: "Review was created successfully."
    else

>       render :new, status: :unprocessable_entity

Reviewer 4: although it works fine, I guess we should encourage usage of `unprocessable_content` (here and in all other places), https://github.com/rack/rack/pull/2209

Author: Agreed, probably need a separate PR to update all the other guides, but I've updated this one.

to avoid N+1 queries.

Let's update `app/views/products/show.html.erb` to also pass along `@reviews` to
the partial so it knows which reviews to render.

```erb
<%= render "reviews", product: @product, reviews: @reviews %>
```

### Calcuating rating percentages

Let's add a method to calculate the percentage of reviews at a specific rating.
Given a rating of `4`, this will return the percentage of reviews with a 4 star rating.

Add the following to `app/models/product.rb`:

```ruby#4-6
class Product < ApplicationRecord
  # ...

  def rating_percentage(rating)
    (reviews.where(rating: rating).count.to_f / reviews_count * 100).round
  end
end
```

We can use this to display each rating and their percentage of reviews with that
rating.

### Product reviews layout

Let's update `app/views/products/_reviews.html.erb` to display that information
alongside the reviews.

```erb
<section class="reviews">
  <aside>
    <h3>Reviews</h3>

    <% if product.reviews_count > 0 %>
      <div>
        <% 5.times do |i| %>
          <%= tag.span "★", class: (i < product.rating.round ? "gold" : "gray") %>
        <% end %>
        <%= product.rating %> out of 5

>       </div>

Reviewer 2: ```suggestion
      <div role="img" aria-label="<%= review.rating %> out of 5 stars">
        <% 5.times do |i| %>
          <%= tag.span "★", class: (i < review.rating ? "gold" : "gray"), aria: { hidden: true } %>
        <% end %>
        <%= product.rating %> out of 5
      </div> 
```

      <div><%= pluralize product.reviews_count, "review" %></div>

> 
>       <div>
>         <% 5.downto(1).each do |i| %>
>           <%= link_to product_path(product, rating: i), class: "review__summary" do %>
>             <%= tag.div i, class: "review__stars" %>
>             <div class="gold">★</div>
>             <div class="review__bars">
>               <div class="review__bar--background"></div>
>               <div class="review__bar" style="width: <%= product.rating_percentage(i) %>%;"></div>
>             </div>
>             <div class="review__percentage"><%= product.rating_percentage(i) %>%</div>
>           <% end %>
>         <% end %>

Reviewer 2: ```suggestion
        <% 5.downto(1).each do |i| %>
          <%= link_to product_path(product, rating: i), class: "review__summary", aria: { label: "#{pluralize(i, "star")} — #{product.rating_percentage(i)}% of reviews" } do %>
          <div aria-hidden="true">
            <div class="review__stars"><%= i %></div>
            <div class="gold">★</div>
            <div class="review__bars">
              <div class="review__bar--background"></div>
              <div class="review__bar" style="width: <%= product.rating_percentage(i) %>%;"></div>
            </div>
            <div class="review__percentage"><%= product.rating_percentage(i) %>%</div>
          </div>
        <% end %>
      <% end %>
```

Add a synthetic `aria-label` on the link, and hide everything inside from screen readers.

ratings in the sidebar.

Add the following to `app/assets/stylesheets/application.css`:

```css
section.reviews {
  display: grid;
  grid-template-columns: 250px 1fr;
  margin-top: 2rem;
  gap: 2rem;
}

.review__summary {
  display: flex;
  align-items: center;
  margin-top: 0.25rem;
}

.review__stars {
  width: 1rem;
}

.review__bars {
  position: relative;
  flex: 1;
  display: flex;
  margin: 0px 5px;
}
.review__bar--background {
  background:#eee;
  border-radius: 15px;
  flex: 1;
  height: 12px;
}
.review__bar {
  background: gold;
  border-radius: 15px;
  position: absolute;
  inset-block: 0;
}

.review__percentage {
  text-align: right;
  width: 2.5rem;
}

.review {
  padding: 1rem;
}

.review__images {
  margin-top: 1em;
  display: flex;
  flex-wrap: wrap;
  gap: 0.5rem;
}
```

### Reviews partial

Let's add `app/views/reviews/_review.html.erb` so the individual reviews can be
displayed.

```erb
<%= tag.div id: dom_id(review), class: "review" do %>
  <div><%= tag.strong review.user.full_name %></div>
  <div>
    <% 5.times do |i| %>
      <%= tag.span "★", class: (i < review.rating ? "gold" : "gray") %>
    <% end %>

>   </div>

Reviewer 2: ```suggestion
  <div role="img" aria-label="<%= review.rating %> out of 5 stars">
    <% 5.times do |i| %>
      <%= tag.span "★", class: (i < review.rating ? "gold" : "gray"), aria: { hidden: true } %>
    <% end %>
  </div>
```

Preserve the visual style while giving screen readers a clear indication of the user rating.

  <%= review.body %>

  <% if review.images.attached? %>
    <div class="review__images">
      <% review.images.each do |image| %>

>         <%= link_to image_tag(image.variant(resize_to_limit: [150, 150])), image, target: :_blank %>

Reviewer 2: ```suggestion
        <%= link_to image, target: :_blank, title: "View full-size image (opens in new tab)", aria: { label: "View full-size image (opens in new tab)" } do %>
          <%= image_tag image.variant(resize_to_limit: [150, 150]), alt: "" %>
        <% end %>
```

Add missing `alt` to image tag, and both `title/aria-label` indicating what happens when following the link.

```ruby#6
class Review < ApplicationRecord
  belongs_to :product, counter_cache: true
  belongs_to :user
  has_many_attached :images

  scope :rated, ->(rating) { rating.present? ? where(rating: rating.to_i) : all }

  validates :body, :rating, presence: true

  after_commit :update_product_rating

  def update_product_rating
    product.update_column(:rating, product.reviews.average(:rating)&.round(1))
  end
end
```

This scope accepts a `rating` argument and filters the query to reviews matching
this rating. If nil or an empty string was passed, it will return all reviews.

The scope also handles invalid ratings safely. For example, `rated("foo")` will
call `"foo".to_i` which returns `0` and will return reviews with a rating of 0.
Since `0` is not a valid rating, there will be no reviews returned.

Let's update `app/controllers/products_controller.rb` to use this new scope.

```ruby#10
class ProductsController < ApplicationController
  allow_unauthenticated_access

  def index
    @products = Product.all
  end

  def show
    @product = Product.find(params[:id])
    @reviews = @product.reviews.with_attached_images.rated(params[:rating])
  end
end
```

> Passing `params[:rating]` let's users filter reviews with a query param. There

Reviewer 2: ```suggestion
Passing `params[:rating]` lets users filter reviews with a query param. There
```

the "Clear filter" link to go back.

## Managing reviews

Admins need the ability to delete spam reviews, fix typos, and correct other
mistakes. Let's build that next.

To start, let's add a resources route in `config/routes.rb` in the store
namespace for reviews.

```ruby#4
# Admins Only
namespace :store do
  resources :products
  resources :reviews
  resources :users
  resources :wishlists
  resources :subscribers
end
```

We can then implement the controller for this in
`app/controllers/store/reviews_controller.rb`:

```ruby
class Store::ReviewsController < Store::BaseController
  before_action :set_review, except: [ :index ]

  def index
    @reviews = Review.includes(:product, :user).with_attached_images.filter_by(params)
  end

  def show
  end

  def edit
  end

  def update
    if @review.update(review_params)
      redirect_to store_review_path(@review)
    else
      render :edit, status: :unprocessable_content
    end
  end

  def destroy
    @review.destroy
    redirect_to store_reviews_path
  end

  private

  def set_review
    @review = Review.find(params[:id])
  end

  def review_params
    params.expect(review: [ :rating, :body, images: [] ])
  end
end
```

> To allow the index to filter by ratings, products, and users, let's a

Reviewer 2: ```suggestion
To allow the index to filter by ratings, products, and users, let's add a
```

    product.update_column(:rating, product.reviews.average(:rating)&.round(1))
  end
end
```

### Sidebar link

To access reviews in the navigation, let's add a link to the sidebar in the
layout.

Add the following to `app/views/layouts/settings.html.erb`:

```erb#13
<%= content_for :content do %>
  <section class="settings">
    <nav>
      <h4>Account Settings</h4>
      <%= link_to "Profile", settings_profile_path %>
      <%= link_to "Email", settings_email_path %>
      <%= link_to "Password", settings_password_path %>
      <%= link_to "Account", settings_user_path %>

      <% if Current.user.admin? %>
        <h4>Store Settings</h4>
        <%= link_to "Products", store_products_path %>
        <%= link_to "Reviews", store_reviews_path %>
        <%= link_to "Users", store_users_path %>
        <%= link_to "Subscribers", store_subscribers_path %>
        <%= link_to "Wishlists", store_wishlists_path %>
      <% end %>
    </nav>

    <div>
      <%= yield %>
    </div>
  </section>
<% end %>

<%= render template: "layouts/application" %>

```

Let's also add a link to the product's show page in `app/views/store/products/show.html.erb`:

```erb#4
<%# ... %>

<section>

>   <%= link_to pluralize(@product.reviews_count), "review"), store_reviews_path(product_id: @product.id) %>

Reviewer 5: There is an unbalanced parenthesis, I think this should be `pluralize(@product.reviews_count, "review")`. Same issue on the wishlists and subscribers lines just below.

Btw, this is for another discussion but I think there is also a bug on the Rails helper, I couldn't find a way to use it as expected:

<img width="351" height="177" alt="Image" src="https://github.com/user-attachments/assets/125537b8-6325-4994-bb03-4238da19db0d" />

Tested on:

- ruby 3.3.11
- Rails 7.2.3.1

Author: Thanks, fixed!

```erb
<h1>Reviews</h1>

<%= form_with url: store_reviews_path, method: :get do |form| %>
  <%= form.collection_select :product_id, Product.all, :id, :name, selected: params[:product_id], include_blank: "All Products" %>
  <%= form.select :rating, 5.downto(1).map{ pluralize it, "star" }, selected: params[:rating], include_blank: "All Ratings" %>
  <%= form.collection_select :user_id, User.all, :id, :full_name, selected: params[:user_id], include_blank: "All Users" %>
  <%= form.submit "Filter" %>
<% end %>

<%= render @reviews %>
```

Next, let's create the reviews partial in the store namespace. This will look
very similar to the public version, but with some additional context since we're
viewing reviews for any product.

Create `app/views/store/reviews/_review.html.erb` with the following:

```erb
<%= tag.div id: dom_id(review), class: "review" do %>
  <div><%= link_to review.user.full_name, store_user_path(review.user) %> reviewed <%= link_to review.product.name, store_product_path(review.product) %></div>
  <div>
    <% 5.times do |i| %>
      <%= tag.span "★", class: (i < review.rating.round ? "gold" : "gray") %>
    <% end %>

>   </div>

Reviewer 2: ```suggestion
  <div role="img" aria-label="<%= review.rating %> out of 5 stars">
    <% 5.times do |i| %>
      <%= tag.span "★", class: (i < review.rating ? "gold" : "gray"), aria: { hidden: true } %>
    <% end %>
  </div>
```

Same as before, hide the stars from screen readers and render a clear indication instead.

  <%= review.body %>

  <% if review.images.attached? %>
    <div class="review__images">
      <% review.images.each do |image| %>

>         <%= link_to image_tag(image.variant(resize_to_limit: [150, 150])), image, target: :_blank %>

Reviewer 2: ```suggestion
        <%= link_to image, target: :_blank, title: "View full-size image (opens in new tab)", aria: { label: "View full-size image (opens in new tab)" } do %>
          <%= image_tag image.variant(resize_to_limit: [150, 150]), alt: "" %>
        <% end %>
```

Same as before: add missing `alt` attribute, and both `title/aria-label` to indicate what happens when following the link.

      <% end %>
    </div>
  <% end %>

  <div>
    <%= link_to "View review", store_review_path(review) %>
  </div>
<% end %>
```

### Show view

Let's create the show view next in `app/views/store/reviews/show.html.erb`:

```erb
<%= link_to "Back to all reviews", store_reviews_path %>

<h1>Review</h1>

> <%= tag.div id: dom_id(@review), class: "review" do %>
>   <div><%= link_to @review.user.full_name, store_user_path(@review.user) %> reviewed <%= link_to @review.product.name, store_product_path(@review.product) %></div>
>   <div>
>     <% 5.times do |i| %>
>       <%= tag.span "★", class: (i < @review.rating.round ? "gold" : "gray") %>
>     <% end %>
>   </div>
> 
>   <%= @review.body %>
> 
>   <% if @review.images.attached? %>
>     <div class="review__images">
>       <% @review.images.each do |image| %>
>         <%= link_to image_tag(image.variant(resize_to_limit: [150, 150])), image, target: :_blank %>
>       <% end %>
>     </div>
>   <% end %>
> <% end %>

Reviewer 2: ```suggestion
<%= render "store/reviews/review", review: @review %>
```

<div>
  <%= link_to "Edit", edit_store_review_path(@review) %>
  <%= button_to "Delete", store_review_path(@review), method: :delete, data: {turbo_confirm: "Are you sure?"} %>
</div>
```

### Edit view

Last, but not least, let's create the edit view in `app/views/store/reviews/edit.html.erb`

```erb
<h1>Edit Review</h1>

<%= form_with model: [:store, @review] do |form| %>
  <div>
    <%= form.label :rating %>
    <div class="rating">
      <% 5.downto(1).each do |i| %>
        <%= form.radio_button :rating, i, required: true %>
        <%= form.label :rating, "★", value: i %>
      <% end %>

>     </div>

Reviewer 2: ```suggestion
  <fieldset>
    <legend>Rating</legend>
    <div class="rating">
      <% 1.upto(5).each do |i| %>
        <%= form.radio_button :rating, i, required: true, class: "sr-only" %>
        <%= form.label :rating, value: i do %>
          <span aria-hidden="true">★</span>
          <span class="sr-only"><%= pluralize(i, "star") %></span>
        <% end %>
      <% end %>
    </div>
  </fieldset>
```

Running 8 tests in a single process (parallelization threshold is 50)
Run options: --seed 41516

# Running:

........

Finished in 0.689592s, 11.6011 runs/s, 18.8517 assertions/s.
8 runs, 13 assertions, 0 failures, 0 errors, 0 skips
```

Great!

We should also run the entire test suite to ensure everything passes.

```bash
$ bin/rails test
Running 40 tests in a single process (parallelization threshold is 50)
Run options: --seed 27614

# Running:

........................................

Finished in 1.865963s, 21.4367 runs/s, 55.7353 assertions/s.
40 runs, 104 assertions, 0 failures, 0 errors, 0 skips
```

## Deploying to Production

Now that we're finished adding product reviews, let's deploy them to production.
Commit and push your changes to the Git repository and then run:

```bash
$ bin/kamal deploy
```

## What's Next

Your e-commerce store now has product reviews to help customers make more
informed decisions about your products!

Here are a few ideas to build on to this:

- Write more tests
- Finish translating the app into another language
- Add a carousel for product images
- Improve the design with CSS
- Add payments to buy products

> 

Reviewer 2: ```suggestion
- Limit the number of reviews displayed on the product page

```

to avoid N+1 queries.

Let's update `app/views/products/show.html.erb` to also pass along `@reviews` to
the partial so it knows which reviews to render.

```erb
<%= render "reviews", product: @product, reviews: @reviews %>
```

### Calcuating rating percentages

Let's add a method to calculate the percentage of reviews at a specific rating.
Given a rating of `4`, this will return the percentage of reviews with a 4 star rating.

Add the following to `app/models/product.rb`:

```ruby#4-6
class Product < ApplicationRecord
  # ...

  def rating_percentage(rating)
    (reviews.where(rating: rating).count.to_f / reviews_count * 100).round
  end
end
```

We can use this to display each rating and their percentage of reviews with that
rating.

### Product reviews layout

Let's update `app/views/products/_reviews.html.erb` to display that information
alongside the reviews.

```erb
<section class="reviews">
  <aside>
    <h3>Reviews</h3>

    <% if product.reviews_count > 0 %>
      <div>
        <% 5.times do |i| %>
          <%= tag.span "★", class: (i < product.rating.round ? "gold" : "gray") %>
        <% end %>
        <%= product.rating %> out of 5

>       </div>

Author: ```suggestion
      <div role="img" aria-label="<%= product.rating.round %> out of 5 stars">
        <% 5.times do |i| %>
          <%= tag.span "★", class: (i < product.rating.round ? "gold" : "gray"), aria: { hidden: true } %>
        <% end %>
        <%= product.rating %> out of 5
      </div>
```

- `user:belongs_to` associates the `Review` with the `User` who created it
- `rating` is an integer that stores the 1-5 star rating
- `body` stores the text description of the review
- `images` stores images using ActiveStorage

### Caches

Before we run this migration, let's modify it to add a couple things to the
`products` table that we'll need.

1. A counter cache to keep track of a product's total number of reviews
2. A rating column to store the product's average rating

Open `db/migrate/<timestamp>_create_reviews.rb` and add these columns:

```ruby#12-13
class CreateReviews < ActiveRecord::Migration[8.0]
  def change
    create_table :reviews do |t|
      t.belongs_to :product, null: false, foreign_key: true
      t.belongs_to :user, null: false, foreign_key: true
      t.integer :rating, null: false
      t.text :body, null: false

      t.timestamps
    end

    add_column :products, :reviews_count, :integer, default: 0
    add_column :products, :rating, :decimal, precision: 2, scale: 1, default: 0
  end
end
```

We used `decimal` as the type for `rating` with a couple options to control
how the numbers are stored.

- `precision` is the total number of digits

> - `scale` is the number of digitals after the decimal

Reviewer 6: Probably a typo, but I think we should use `digits` for both `precision` and `scale`:
```suggestion
- `scale` is the number of digits after the decimal
```

This means we can store up to `9.9`. Two digits total with one after the decimal

> place.

Reviewer 6: This sentence was a bit confusing to me, maybe something like:
```suggestion
This means we can store up to `9.9`. Two digits in total, with one digit after the
decimal sign.
```

In the terminal, run the migrations to update the database.

```bash
$ bin/rails db:migrate
== 20260421200530 CreateReviews: migrating ====================================
-- create_table(:reviews)
   -> 0.0036s
-- add_column(:products, :reviews_count, :integer, {default: 0})
   -> 0.0005s
-- add_column(:products, :rating, :decimal, {precision: 2, scale: 1})
   -> 0.0004s
== 20260421200530 CreateReviews: migrated (0.0045s) ===========================
```

> Next, let's validate that every review has a body and rating in `app/models/review.rb`:

Reviewer 6: To me it was somewhat confusing, why the `belongs_to :product` line is highlighted, when we are only adding validations. It is explained after the example, but maybe we can mention it here as well?
Another option would be a separate example that only adds the counter_cache, but maybe that would be overkill?
Something like:
```suggestion
Next, let's update the model and validate that every review has a body and rating in `app/models/review.rb`:
```

<%= form_with model: [@product, @review] do |form| %>
  <fieldset>
    <legend>Rating</legend>
    <div class="rating">
      <% 1.upto(5).each do |i| %>
        <%= form.radio_button :rating, i, required: true, class: "sr-only" %>
        <%= form.label :rating, value: i do %>
          <span aria-hidden="true">★</span>
          <span class="sr-only"><%= pluralize(i, "star") %></span>
        <% end %>
      <% end %>
    </div>
  </fieldset>

  <div>
    <%= form.label :body, style: "display: block;" %>
    <%= form.textarea :body, required: true %>
  </div>

  <div>
    <%= form.label :images, style: "display: block;" %>
    <%= form.file_field :images, multiple: true, accept: "image/*", capture: "environment" %>
  </div>

  <%= form.submit %>
<% end %>
```

We're using a loop to create 5 radio buttons for the different ratings a user
could choose.

For image uploads, we're using a file field with a couple attributes that do the
following:

- `multiple: true` tells the browser to allow multiple file uploads
- `accept: "image/*"` is a browser hint that filters the file selector to only
images
- `capture: "environment"` is used by mobile browsers to enable the camera using

> the rear camera

Reviewer 6: Is it common to use the term "rear camera"?
Mozilla uses [outward-facing](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/capture) which is a bit clearer to me as a non-native speaker.
Maybe:
```suggestion
the outward-facing camera
```

> ### Styling the star rating input

Reviewer 6: Could you update all the headings to capitalize all words except for prepositions, conjunctions, internal articles, and forms of the verb "to be"?
https://edgeguides.rubyonrails.org/ruby_on_rails_guides_guidelines.html#headings
```suggestion
### Styling the Star Rating Input
```

The stars should be gray by default and turn gold when you select a rating. We
can use a few CSS tricks to make this work like you'd expect.

Add the following to `app/assets/stylesheets/application.css` at the bottom:

```css
/* Remove default styling for fieldset */
fieldset {
  border: 0;
  padding: 0;
  margin: 0;
}

/* Remove default styling for legend */
legend {
  padding: 0;
}

/* Hide text visually but not from screen readers */
.sr-only {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}

.rating {
  display: flex;
}

> /* Lighten all stars by default */

Reviewer 6: Nitpick. Lighten to me suggests highlight/brighten, where the opposite is probably what we want.
Maybe (or de-emphasize, gray-out) :
```suggestion
/* Mute all stars by default */
```

.rating label {
  color: lightgray;
}

> /* Gold: the selected/focused/hovered star and all previous ones */

Reviewer 6: It would be nice if this comment is more similar to the other comments.
For example:
```suggestion
/* Highlight the selected/focused/hovered star and all previous ones */
```

.rating input:is(:checked, :focus) + label,
.rating label:has(~ input:is(:checked, :focus)),
.rating label:hover,
.rating label:has(~ input + label:hover) {
  color: gold;
}

> /* Visible focus indicator */

Reviewer 6: Maybe this comment could be more descriptive?
```suggestion
/* Outline visible focused inputs */
```

```ruby#10
class ProductsController < ApplicationController
  allow_unauthenticated_access

  def index
    @products = Product.all
  end

  def show
    @product = Product.find(params[:id])
    @reviews = @product.reviews.with_attached_images.rated(params[:rating])
  end
end
```

Passing `params[:rating]` lets users filter reviews with a query param. There
are several things that can happen:

- If the URL contains `?rating=3`, this will only return reviews with a rating
of 3.
- If this `rating` param is empty or missing, it will return all reviews.
- If the rating param has an invalid value like `?rating=foo`, it will return
no reviews.

Let's update `app/views/products/_reviews.html.erb` to display the filter and a
way to clear it.

```erb#7-13
<section class="reviews">
  <aside>
    <%# ... %>
  </aside>

  <div>
    <% if params[:rating] %>
      <div>
        Filtered by <%= pluralize params[:rating].to_i, "star" %>.
        <%= link_to "Clear filter", product %>
      </div>
    <% end %>

    <%= render reviews %>
  </div>
</section>
```

Test out filtering by clicking on a rating on the left to filter reviews. Use

> the "Clear filter" link to go back.

Reviewer 6: Nitpick. "go back" to me implies the browser going back in history.
Maybe:
```suggestion
the "Clear filter" link to show unfiltered reviews.
```
Or maybe:
```suggestion
the "Clear filter" link to show all reviews.
```

  <div>
    <%= form.label :body, style: "display: block;" %>
    <%= form.textarea :body, required: true %>
  </div>

  <div>
    <%= form.label :images, style: "display: block;" %>
    <%= form.file_field :images, multiple: true, accept: "image/*", capture: "environment" %>

    <% form.object.images.each do |image| %>
      <div>
        <%= image_tag image.variant(resize_to_limit: [150, 150]) %>
        <%= form.hidden_field :images, value: image.signed_id, multiple: true, id: nil %>
        <button onclick="this.parentElement.remove()">Remove</button>
      </div>
    <% end %>
  </div>

  <%= form.submit %>
<% end %>
```

By default, Rails will replace all of the existing images when a new value is
assigned. To preserve existing images when editing a review, we need to create
hidden fields for each image that's already uploaded. These hidden fields use
the image's `signed_id` to reference the existing ActiveRecord object and also
ensures that it wasn't tampered with. We use `multiple: true` so Rails generates
the correct name to add the `images` param as an array. We also disable the `id`
generated for this fields since it would generate duplicates and we don't need
them.

> With that added, store admins can now view, edit and delete reviews as needed.

Reviewer 6: ```suggestion
With that added, store admins can now view, edit, and delete reviews as needed.
```
